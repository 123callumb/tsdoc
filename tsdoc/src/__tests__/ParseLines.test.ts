import { TSDocParser, DocComment } from '../index';

function escape(s: string): string {
  return s.replace(/\n/g, '[n]')
    .replace(/\r/g, '[r]');
}

function parseAndMatchSnapshot(buffer: string): void {
  const tsdocParser: TSDocParser = new TSDocParser();
  const docComment: DocComment = tsdocParser.parseString(buffer);
  expect({
    buffer: escape(buffer),
    errors: docComment.parseErrors,
    comment: escape(docComment.range.toString()),
    lines: docComment.lines.map(line => escape(line.toString()))
  }).toMatchSnapshot();
}

test('Newline examples', () => {
  parseAndMatchSnapshot([
    '',
    '/**',
    ' * L1',
    ' */',
    ''
  ].join('\r\n'));
  parseAndMatchSnapshot([
    '/**',
    'L1',
    'L2',
    '*/'
  ].join('\r\n'));

  // We currently don't support CR or LFCR, so a single "\r" is treated
  // as part of the line.
  parseAndMatchSnapshot(`/** L \r 1 */`);
});

test('Spacing variations', () => {
  parseAndMatchSnapshot(`/***/`);                      // 1
  parseAndMatchSnapshot(` /***/ `);                    // 2
  parseAndMatchSnapshot(` /** */ `);                   // 3
  parseAndMatchSnapshot(` /**\n\n*/ `);                // 4
  parseAndMatchSnapshot(` /**L1*/ `);                  // 5
  parseAndMatchSnapshot(` /** L1 */ `);                // 6
  parseAndMatchSnapshot(` /**L1\n*/ `);                // 7
  parseAndMatchSnapshot(` /**L1*\n*/ `);               // 8
  parseAndMatchSnapshot(` /**\nL1*/ `);                // 9
  parseAndMatchSnapshot(` /**\n L1 */ `);              // 10
  parseAndMatchSnapshot(` /**\nL1\n*/ `);              // 11
  parseAndMatchSnapshot(` /**\nL1\n\nL2*/ `);          // 12
  parseAndMatchSnapshot(` /**\n*L1\n*/ `);             // 13
  parseAndMatchSnapshot(` /**\n * L1\n*/ `);           // 14
  parseAndMatchSnapshot(` /**\n * L1\n */ `);          // 15
  parseAndMatchSnapshot(` /**L1\n *L2\nL3*/ `);        // 16
  parseAndMatchSnapshot(` /** L1\n * L2\n L3*/ `);     // 17
  parseAndMatchSnapshot(` /** L1 \n * L2 \n L3 */ `);  // 18
  parseAndMatchSnapshot([                              // 19
    '/**  L1  ',
    ' *  L2  ',
    '  L3  ',
    '  L4 */'
  ].join('\r\n'));
});

// TODO: Special handling for these somewhat common ornamentations
test('Stars added', () => {
  parseAndMatchSnapshot(` /****/ `);
  parseAndMatchSnapshot(` /**L1**/ `);
  parseAndMatchSnapshot(` /***L1*/ `);
  parseAndMatchSnapshot(`
/*****
 **X**
 *****/ `);
});
