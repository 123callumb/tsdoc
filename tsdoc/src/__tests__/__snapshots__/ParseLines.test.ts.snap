// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newline examples 1`] = `
Object {
  "buffer": "[r][n]/**[r][n] * L1[r][n] */[r][n]",
  "comment": "/**[r][n] * L1[r][n] */",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Newline examples 2`] = `
Object {
  "buffer": "/**[r][n]L1[r][n]L2[r][n]*/",
  "comment": "/**[r][n]L1[r][n]L2[r][n]*/",
  "errors": Array [],
  "lines": Array [
    "L1",
    "L2",
  ],
}
`;

exports[`Newline examples 3`] = `
Object {
  "buffer": "/** L [r] 1 */",
  "comment": "/** L [r] 1 */",
  "errors": Array [],
  "lines": Array [
    "L [r] 1",
  ],
}
`;

exports[`Spacing variations 1`] = `
Object {
  "buffer": "/***/",
  "comment": "/***/",
  "errors": Array [],
  "lines": Array [],
}
`;

exports[`Spacing variations 2`] = `
Object {
  "buffer": " /***/ ",
  "comment": "/***/",
  "errors": Array [],
  "lines": Array [],
}
`;

exports[`Spacing variations 3`] = `
Object {
  "buffer": " /** */ ",
  "comment": "/** */",
  "errors": Array [],
  "lines": Array [],
}
`;

exports[`Spacing variations 4`] = `
Object {
  "buffer": " /**[n][n]*/ ",
  "comment": "/**[n][n]*/",
  "errors": Array [],
  "lines": Array [
    "",
  ],
}
`;

exports[`Spacing variations 5`] = `
Object {
  "buffer": " /**L1*/ ",
  "comment": "/**L1*/",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Spacing variations 6`] = `
Object {
  "buffer": " /** L1 */ ",
  "comment": "/** L1 */",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Spacing variations 7`] = `
Object {
  "buffer": " /**L1[n]*/ ",
  "comment": "/**L1[n]*/",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Spacing variations 8`] = `
Object {
  "buffer": " /**L1*[n]*/ ",
  "comment": "/**L1*[n]*/",
  "errors": Array [],
  "lines": Array [
    "L1*",
  ],
}
`;

exports[`Spacing variations 9`] = `
Object {
  "buffer": " /**[n]L1*/ ",
  "comment": "/**[n]L1*/",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Spacing variations 10`] = `
Object {
  "buffer": " /**[n] L1 */ ",
  "comment": "/**[n] L1 */",
  "errors": Array [],
  "lines": Array [
    " L1",
  ],
}
`;

exports[`Spacing variations 11`] = `
Object {
  "buffer": " /**[n]L1[n]*/ ",
  "comment": "/**[n]L1[n]*/",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Spacing variations 12`] = `
Object {
  "buffer": " /**[n]L1[n][n]L2*/ ",
  "comment": "/**[n]L1[n][n]L2*/",
  "errors": Array [],
  "lines": Array [
    "L1",
    "",
    "L2",
  ],
}
`;

exports[`Spacing variations 13`] = `
Object {
  "buffer": " /**[n]*L1[n]*/ ",
  "comment": "/**[n]*L1[n]*/",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Spacing variations 14`] = `
Object {
  "buffer": " /**[n] * L1[n]*/ ",
  "comment": "/**[n] * L1[n]*/",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Spacing variations 15`] = `
Object {
  "buffer": " /**[n] * L1[n] */ ",
  "comment": "/**[n] * L1[n] */",
  "errors": Array [],
  "lines": Array [
    "L1",
  ],
}
`;

exports[`Spacing variations 16`] = `
Object {
  "buffer": " /**L1[n] *L2[n]L3*/ ",
  "comment": "/**L1[n] *L2[n]L3*/",
  "errors": Array [],
  "lines": Array [
    "L1",
    "L2",
    "L3",
  ],
}
`;

exports[`Spacing variations 17`] = `
Object {
  "buffer": " /** L1[n] * L2[n] L3*/ ",
  "comment": "/** L1[n] * L2[n] L3*/",
  "errors": Array [],
  "lines": Array [
    "L1",
    "L2",
    " L3",
  ],
}
`;

exports[`Spacing variations 18`] = `
Object {
  "buffer": " /** L1 [n] * L2 [n] L3 */ ",
  "comment": "/** L1 [n] * L2 [n] L3 */",
  "errors": Array [],
  "lines": Array [
    "L1",
    "L2",
    " L3",
  ],
}
`;

exports[`Spacing variations 19`] = `
Object {
  "buffer": "/**  L1  [r][n] *  L2  [r][n]  L3  [r][n]  L4 */",
  "comment": "/**  L1  [r][n] *  L2  [r][n]  L3  [r][n]  L4 */",
  "errors": Array [],
  "lines": Array [
    " L1",
    " L2",
    "  L3",
    "  L4",
  ],
}
`;

exports[`Stars added 1`] = `
Object {
  "buffer": " /****/ ",
  "comment": "/****/",
  "errors": Array [],
  "lines": Array [
    "*",
  ],
}
`;

exports[`Stars added 2`] = `
Object {
  "buffer": " /**L1**/ ",
  "comment": "/**L1**/",
  "errors": Array [],
  "lines": Array [
    "L1*",
  ],
}
`;

exports[`Stars added 3`] = `
Object {
  "buffer": " /***L1*/ ",
  "comment": "/***L1*/",
  "errors": Array [],
  "lines": Array [
    "*L1",
  ],
}
`;

exports[`Stars added 4`] = `
Object {
  "buffer": "[n]/*****[n] **X**[n] *****/ ",
  "comment": "/*****[n] **X**[n] *****/",
  "errors": Array [],
  "lines": Array [
    "***",
    "*X**",
    "***",
  ],
}
`;
